astropy/cosmology/io/yaml.py

```diff
@@ -121,20 +121,38 @@ def register_cosmology_yaml(cosmo_cls):
 # Unified-I/O Functions
 
 
-def from_yaml(yml):
+def from_yaml(yml, *, cosmology=None):
     """Load `~astropy.cosmology.Cosmology` from :mod:`yaml` object.
 
     Parameters
     ----------
     yml : str
         :mod:`yaml` representation of |Cosmology| object
+    cosmology : str, `~astropy.cosmology.Cosmology` class, or None (optional, keyword-only)
+        The expected cosmology class (or string name thereof). This argument is
+        is only checked for correctness if not `None`.
 
     Returns
     -------
     `~astropy.cosmology.Cosmology` subclass instance
+
+    Raises
+    ------
+    TypeError
+        If the |Cosmology| object loaded from ``yml`` is not an instance of
+        the ``cosmology`` (and ``cosmology`` is not `None`).
     """
     with u.add_enabled_units(cu):
-        return load(yml)
+        cosmo = load(yml)
+
+    # Check argument `cosmology`, if not None
+    # This kwarg is required for compatibility with |Cosmology.from_format|
+    if isinstance(cosmology, str):
```

Comment:
This is not very fool-proof, e.g., `cosmology=[]` will give a strange error. I personally do not care too much about wrong input, but perhaps there is a case to factor getting the class out into something more like what is done for coordinate frames: https://github.com/astropy/astropy/blob/41f023995fc2b533deeac84f6fa818a5df04c26e/astropy/coordinates/baseframe.py#L38-L51

Reply:
I added a card to https://github.com/orgs/astropy/projects/8